microceph/database/client_config_extras.go

```diff
@@ -0,0 +1,220 @@
+	return ccq.GetAllForFilter(s, ClientConfigItemFilter{Host: nil, Key: &key})
+}
+
+func (ccq ClientConfigQueryImpl) GetAllForHost(s *state.State, host string) (ClientConfigItems, error) {
+	return ccq.GetAllForFilter(s, ClientConfigItemFilter{Host: &host, Key: nil})
+}
+
+func (ccq ClientConfigQueryImpl) GetAllForKeyAndHost(s *state.State, key string, host string) (ClientConfigItems, error) {
+	return ccq.GetAllForFilter(s, ClientConfigItemFilter{Host: &host, Key: &key})
+}
+
+func (ccq ClientConfigQueryImpl) GetAllForFilter(s *state.State, filters ...ClientConfigItemFilter) (ClientConfigItems, error) {
+	var err error
+	var retval []ClientConfigItem
+
+	err = s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		retval, err = GetClientConfigItems(ctx, tx, filters...)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return retval, err
+	}
+	return retval, nil
+}
+
+// Delete Methods
+func (ccq ClientConfigQueryImpl) RemoveAllForKey(s *state.State, key string) error {
+	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		err := DeleteClientConfigItems(ctx, tx, key)
+		if err != nil {
+			return fmt.Errorf("failed to clean existing keys %s: %v", key, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ccq ClientConfigQueryImpl) RemoveOneForKeyAndHost(s *state.State, key string, host string) error {
+	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		err := DeleteClientConfigItem(ctx, tx, key, host)
+		if err != nil {
+			return fmt.Errorf("failed to clean existing keys %s: %v", key, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Internal Methods
+func addNewGlobalClientConfig(s *state.State, key string, value string) error {
+	members, err := common.GetClusterMemberNames(common.CephState{State: s})
+	if err != nil {
+		return err
+	}
+
+	err = s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		// Delete all instances of key before adding the global config.
+		err := DeleteClientConfigItems(ctx, tx, key)
+		if err != nil {
+			return fmt.Errorf("failed to clean existing keys %s: %v", key, err)
+		}
+
+		for _, member := range members {
+			// Populate record for the member
+			logger.Infof("Host is: %v", member)
+
+			// Add record to database.
+			_, err = CreateClientConfigItem(ctx, tx, ClientConfigItem{Key: key, Value: value, Host: member})
+			if err != nil {
+				return fmt.Errorf("failed to add client config: %v", err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func addNewHostClientConfig(s *state.State, key string, value string, host string) error {
+	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		// Remove existing client config entry.
+		err := DeleteClientConfigItem(ctx, tx, key, host)
```

Comment:
Hm, I wonder if we need to delete first? Could we just update an existing entry? Sqlite supports `INSERT OR REPLACE` that might be useful here.

Reply:
Insert for both global and host configs have been changed to use `INSERT OR REPLACE` hence delete config prior to create has been removed.